### PR TITLE
Refactored the code with PyTorch Lightning (WIP)

### PR DESCRIPTION
**Pull request by hanaol**
1. Setup PyTorch Lightning integration.
2. Updated the versions in requirements.txt

### CODE REVIEW

File: src/electrai/configs/MP/config.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+save_every_epochs: 2
+epochs: 2
+nbatch: 1
+lr: 0.01
+weight_decay: 0.0
+
+# Weights and biases
+wb_pname: "electrai-test"
+wb_ename: "mp-no-upscaling"
+entity: "hanaoli"
```
Comment (forklady42) [inline src/electrai/configs/MP/config.yaml]: probably better to check in "PrinceOA"

File: src/electrai/entrypoints/train.py [hunk 2]
```diff
@@ -1,176 +1,101 @@
...
-        if t % int(cfg.save_every_epochs) == 1:
-            save_checkpoint(
-                t, model, optimizer, scheduler, f"{cfg.model_prefix}_{t}.pth"
-            )
+    lr_monitor = LearningRateMonitor(logging_interval="epoch")
+
+    # -----------------------------
+    # Profiler
+    # -----------------------------
+    profiler = PyTorchProfiler(
```
Comment (forklady42) [inline src/electrai/entrypoints/train.py]: Let's make the profiler optional rather than always setting it up. Profiling adds overhead, so we should only run it when we explicitly want to profile a run.

File: src/electrai/lightning.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        torch.cuda.empty_cache()
+        return loss
+
+    def validation_step(self, batch):
+        X, y = batch
+        pred = self(X)
+        loss = self.loss_fn(pred, y)
+        self.log("val_loss", loss, prog_bar=True, on_step=True, on_epoch=True)
+        # release GPU cache for next batch
+        torch.cuda.empty_cache()
```
Comment (forklady42) [inline src/electrai/lightning.py]: Same comment on removing `torch.cuda.empty_cache()`.

File: src/electrai/lightning.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        )
+
+        linsch = torch.optim.lr_scheduler.LinearLR(
+            optimizer, start_factor=1e-5, end_factor=1, total_iters=1
+        )
+        cossch = torch.optim.lr_scheduler.CosineAnnealingLR(
+            optimizer, T_max=int(self.cfg.epochs) - 1
+        )
+        scheduler = torch.optim.lr_scheduler.SequentialLR(
+            optimizer, [linsch, cossch], milestones=[1]
```
Comment (forklady42) [inline src/electrai/lightning.py]: Since the number of epochs for the warmup comes up a couple of times here, it would be nice to break it out into its own variable just before `linsch`. It will make the code easier to read and reduce the chance that the warmup length will be increased in one area and not another.

File: src/electrai/dataloader/qm9.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        self.ds_data = downsample_data
+        self.ds_label = downsample_label
+        self.da = data_augmentation
+        self.data = data
+        self.data_precision = data_precision
+
+    def __len__(self):
+        return len(self.data)
+
+    def rotate_x(self, data_in):
```
Comment (forklady42) [inline src/electrai/dataloader/qm9.py]: This doesn't need to happen in this PR but note for the future that there's a lot of duplicated code between dataloaders which should be abstracted.

**Review by forklady42 — CHANGES_REQUESTED** [review]
The crux of the refactor looks good! Left some comments with some small-ish changes. Removing `empty_cache` is the main one that needs to be made before merging.

Not needed to change here, but in the future, I would recommend putting the commit with the CHGCAR files in its own PR so that it doesn't clutter the rest of the changes.

File: src/electrai/lightning.py (hunk 4 above)
Comment (hanaol) [reply src/electrai/lightning.py]: @forklady42 Are you referring to defining a variable for the warmup length and use it as follows?

```
        linsch = torch.optim.lr_scheduler.LinearLR(
            optimizer, start_factor=1e-5, end_factor=1, total_iters=self.cfg.warmup_length
        )
        cossch = torch.optim.lr_scheduler.CosineAnnealingLR(
            optimizer, T_max=int(self.cfg.epochs) - self.cfg.warmup_length
        )
        scheduler = torch.optim.lr_scheduler.SequentialLR(
            optimizer, [linsch, cossch], milestones=[self.cfg.warmup_length]
```

File: src/electrai/lightning.py (hunk 3 above)
Comment (hanaol) [reply src/electrai/lightning.py]: sgtm!

File: src/electrai/entrypoints/train.py [hunk 6]
```diff
@@ -1,176 +1,101 @@
...
 from pathlib import Path
 from types import SimpleNamespace
 
-import numpy as np
 import torch
 import yaml
+from lightning.pytorch.profilers import PyTorchProfiler
+from pytorch_lightning import Trainer
+from pytorch_lightning.callbacks import LearningRateMonitor, ModelCheckpoint
+from pytorch_lightning.loggers import WandbLogger
```
[the remark this answers is not in the document]
Comment (hanaol) [reply src/electrai/entrypoints/train.py]: Good point! I' ll change `train.py` as follows:
```
from lightning.pytorch import Trainer
from lightning.pytorch.callbacks import LearningRateMonitor, ModelCheckpoint
```

also in `lightning.py`:
```
from lightning.pytorch import LightningModule
```

**Review by hanaol — COMMENTED** [review]
@forklady42 I have made several changes based on your comments.

File: src/electrai/lightning.py [hunk 7]
```diff
@@ -26,21 +26,26 @@ def forward(self, x):
...
-        self.log("train_loss", loss, prog_bar=True, on_step=True, on_epoch=True)
-        # release GPU cache for next batch
-        torch.cuda.empty_cache()
+        self.log(
+            "train_loss",
+            loss,
+            prog_bar=True,
+            on_step=True,
+            on_epoch=True,
+            sync_dist=True,
```
Comment (forklady42) [inline src/electrai/lightning.py]: Syncing across GPUs can slow down training significantly as the different GPUs must wait on each other. I agree that's we do want the sync for validation loss, but think we can get away without using it here on training loss.

Comment (hanaol) [reply src/electrai/lightning.py]: Good point! I have disabled it for training.
